### ISSUE TITLE
fix(cache): TrainingLoop is cacheable, so a re-run can skip training entirely

### ISSUE THREAD

**Issue by treeleaves30760** [treeleaves30760/CodefyUI#253]
> 中文摘要：`TrainingLoop` 沒有宣告 `cacheable`，所以繼承了預設的 `True`。只要它的上游全部可快取，第二次執行就會整個命中快取：不訓練、不寫 checkpoint、不記錄 metric，而畫布上的 preset 仍然顯示「完成」。**觸發條件比本 issue 最初填的窄**——shipped 的 CNN-MNIST 圖其實碰不到，因為 `ModelSaver` 被強制只能寫在 `./data` 底下，而 `DatasetNode` 的指紋會遞迴走訪整個 `./data`，每次執行都被弄髒。但那是碰巧的保護：把 saver 拿掉（訓練＋畫 loss 曲線，一個很普通的教學圖），全部用真實節點，第 2、3 次執行 `TrainingLoop` 就都是 `cached`。修法也不是把 `cacheable` 設成 `False` 就好——`SequentialModel` 仍被快取且會交回同一個已被訓練過的模型物件，那會變成「每次執行都接續上次的權重繼續訓練」。真正要決定的是：擁有權重的 `SequentialModel` 該不該像其他每個擁有權重的節點一樣不可快取。

> [!IMPORTANT]
> **This issue was corrected after filing.** The original text claimed the shipped CNN-MNIST example reproduces this; it does not, and the first probe that suggested otherwise had substituted a test-only node for the real `Dataset → DataLoader` chain. The full correction, with the evidence for both the negative and the positive result, is in [this comment](https://github.com/CodefyUI/CodefyUI/issues/253#issuecomment-5221012770). The body below has been rewritten to match. The bug is real and reachable; the trigger condition is narrower than first filed.

Found while fixing #223. #223 named this node under "also worth checking". Filed separately rather than folded into #223's PR because the one-line version of the fix is not correct on its own.

## The gap

`TrainingLoopNode` has **no `cacheable` declaration anywhere in the file**, so it inherits `True` from `BaseNode`. On a cache hit the engine returns the recorded outputs without constructing or calling the node (`backend/app/core/graph_engine.py:2045-2075`), so when it hits:

- no training happens at all
- no periodic checkpoint is written (`training_loop_node.py:1289`)
- no interrupt checkpoint is written (`:1362`)
- no TensorBoard event file or `exec_run_artifacts` row is created (`:667`, `:678`)
- no per-epoch metric is logged (`:768-770`) — the loss chart is empty on a run the user was told succeeded

## When it actually triggers

All five nodes inside the `Training Pipeline` preset are `cacheable = True` (`Dataset`, `DataLoader`, `Loss`, `Optimizer`, `TrainingLoop`), and `SequentialModelNode.define_inputs()` returns `[]` — the architecture is a JSON *param*, so it is a fully cacheable root. Nothing in the chain opts out.

**Trigger condition: nothing writes into the dataset's `data_dir` tree between runs.**

Take the shipped `examples/Usage_Example/CNN-MNIST/TrainCNN-MNIST/graph.json`, expand the preset, and drop the `ModelSaver` and its two `Print` consumers. Everything else is untouched and 100% production nodes. Counting real calls to `TrainingLoopNode.execute()`:

```
RUN 1: TrainingLoop.execute calls = 1   dataset=completed  dataloader=completed  optimizer=completed  loss=completed  train_loop=completed
RUN 2: TrainingLoop.execute calls = 0   dataset=cached     dataloader=cached     optimizer=cached     loss=cached     train_loop=cached
RUN 3: TrainingLoop.execute calls = 0   dataset=cached     dataloader=cached     optimizer=cached     loss=cached     train_loop=cached

total TrainingLoop.execute calls across 3 runs: 1
```

"Train a model and plot the loss curve, without saving it" is an ordinary teaching graph.

The `ExecutionCache` is per WebSocket and is deliberately shared across Run clicks (`backend/app/api/ws_execution.py:144`: "It is lent to every run this socket starts"), and `force_rerun` is `set(changed_nodes)` (`graph_engine.py:1775`), empty when nothing was edited. So this is: open the editor, hit Run, hit Run again.

## Why the shipped example does NOT trigger it

Worth recording so nobody re-derives it. The shipped graph keeps its `ModelSaver`, and:

* `DatasetNode.cache_fingerprint` walks `data_dir` (`./data`) **recursively** over the whole shared tree — its own comment flags this as known over-invalidation.
* `ModelSaverNode` writes `./data/models/model_weights.pt`, inside that tree, and **cannot write anywhere else**: `model_saver_node.py:78` raises `ValueError("Output path must be within the project data directory")`.

So every run dirties the dataset's fingerprint, the dataset misses, and the miss propagates to `TrainingLoop`. Measured:

```
fingerprint A: {'count': 14, 'total_size': 68239450, ...} hash=3f7bca63ef4065bbf6f655f536a18b60
(write ./data/models/model_weights.pt)
fingerprint B: {'count': 14, 'total_size': 66549149, ...} hash=93e900bc20267d82ea744f018d40128d
```

**This protection is accidental, and it is load-bearing for every shipped graph that has a saver.** `DatasetNode.cache_fingerprint`'s own comment names scoping the walk to just that dataset's files as "a real follow-up". If that optimisation lands, it removes exactly this protection and the shipped CNN-MNIST graph starts hitting this bug. Whoever picks it up should land the `TrainingLoop` fix first or at the same time.

## Preset status cannot tell you any of this

`_emit_preset_aware` (`graph_engine.py:1856-1861`) rolls `cached` into the container's done-count and emits **`completed`** when the count is full. A preset whose internals were *all* served from cache still reports `completed` — there is no "container was fully cached" status. Any investigation here has to count `execute()` calls or expand the preset; preset-level status is not evidence either way.

That is also the user-visible harm: the canvas says the Training Pipeline completed, and it did nothing.

## Why this is not a one-line fix

Marking `TrainingLoop` `cacheable = False` on its own gives a *different* wrong answer. `SequentialModel` stays cached and hands back the **same module object**, whose weights the previous run's optimizer already mutated in place, so every Run continues training from where the last one stopped. Measured on a hand-built top-level graph with `TrainingLoopNode.cacheable` **monkeypatched to `False`**, SGD lr=0.5, 3 epochs per run:

```
run 1: model node status=completed  first weights=[0.40887, -0.35599, -0.35673]
run 2: model node status=cached     first weights=[0.40419, -0.35878, -0.34844]
run 3: model node status=cached     first weights=[0.39947, -0.36165, -0.34030]
same model object across runs: True
```

(Those numbers are from the monkeypatched variant, not from unmodified code — they show what the naive fix would produce, not current behaviour.)

## The real question

`docs/docs/usage/running-graphs.md` states the existing rule: "every layer that owns weights (`Linear`, `Conv2d`, `LSTM` and the rest), whose parameters drift as training proceeds" is `cacheable = False` — that is what `StatefulModuleMixin` exists for. `SequentialModelNode` owns weights and is a plain `BaseNode`. It looks like the same omission one level up, and fixing it is probably the prerequisite for fixing `TrainingLoop`.

That is a product decision with a real blast radius (`SequentialModel` non-cacheable means every graph rooted in it re-executes fully on every Run), which is why it needs its own design rather than a one-line patch.

## Acceptance

- Two consecutive Runs of a training graph with no saver, nothing edited between them, either both train from the declared initial weights or the behaviour is a documented, deliberate choice the user can see.
- Whatever is decided is pinned by a test that runs the graph twice through `execute_graph` and asserts on **`execute()` call counts or weights**, not on node status — preset aggregation makes status unusable here.
- A note wherever `DatasetNode.cache_fingerprint`'s scoping follow-up is tracked, so that optimisation does not silently re-open this.

Refs #143, #223, #254.

**Comment by treeleaves30760** [comment]
## Correction: the "reachable from the canvas" claim as filed is wrong

> 中文摘要：我原本寫的「標準訓練圖從頭到尾都可快取」是錯的，因為我在探針裡用一個測試專用節點頂替了真正的 `Dataset → DataLoader`，而那個頂替剛好拿掉了真實圖裡的保護。實際跑 shipped 的 CNN-MNIST 圖，第二次執行 `TrainingLoop.execute()` **有**被呼叫，確實有訓練。原因是 `ModelSaver` 被強制只能寫在 `./data` 底下，而 `DatasetNode` 的指紋會遞迴走訪整個 `./data`，所以每次執行 dataset 的指紋都會變、快取一定 miss，再往下游傳播。但這個保護是「碰巧」而不是結構性的：把 saver 拿掉（訓練＋畫 loss 曲線，一個很普通的教學圖），全部用真實節點，`TrainingLoop` 在第 2、3 次執行就都是 `cached`，三次執行總共只訓練了一次。所以 bug 是真的，只是觸發條件比我寫的窄。preset 外框則與此無關。

@ reviewer asked me to reproduce this independently and could not. They were right to push. Here is what I got wrong, what is actually true, and the evidence for both.

### What I got wrong

The issue says the canonical training graph "is cacheable end to end" and implies the shipped example demonstrates it. **It does not.** My probe substituted a test-only `_PureLoaderProbe` node for the real `Dataset → DataLoader` chain, and that substitution is precisely what removed the protection the real graph has. I should have caught that before filing a P8.

Running the shipped `examples/Usage_Example/CNN-MNIST/TrainCNN-MNIST/graph.json` twice against one shared `ExecutionCache`, counting actual calls to `TrainingLoopNode.execute()` (immune to any status plumbing):

```
TrainingLoopNode.execute() calls after run 1: 1
TrainingLoopNode.execute() calls after run 2: 2
=> run 2 executed the training loop: True
```

Run 2 trains. The claim as filed is not supported by the shipped graph.

### The preset wrapper is NOT the explanation

Worth ruling out explicitly, since it was the natural suspicion. I expanded the preset with `expand_presets()` and ran the expanded nodes at top level, where every internal status is visible. Same outcome — `train_loop` executes on every run. The wrapper changes nothing about the cache decision.

It does *hide* it, though, and that is worth knowing on its own: `_emit_preset_aware` (`graph_engine.py:1856-1861`) rolls `cached` into the container's done-count and emits **`completed`** when the count is full. **A preset whose internals were all served from cache still reports `completed`.** There is no "container was fully cached" status, so preset-level status can never answer this question either way.

### Why the shipped graph is actually protected

The expanded run shows which node misses, stable across three runs:

```
RUN 1: TrainingLoop.execute calls = 1   dataset=completed  dataloader=completed  optimizer=completed  loss=completed  train_loop=completed
RUN 2: TrainingLoop.execute calls = 1   dataset=completed  dataloader=completed  optimizer=cached     loss=cached     train_loop=completed
RUN 3: TrainingLoop.execute calls = 1   dataset=completed  dataloader=completed  optimizer=cached     loss=cached     train_loop=completed
```

`dataset` never hits the cache, and that propagates through `dataloader` to `train_loop`.

The cause is an interaction between two nodes that do not know about each other:

* `DatasetNode.cache_fingerprint` walks `data_dir` (`./data`) **recursively**, over the whole shared tree. Its own comment already flags this as known over-invalidation: "writing to an unrelated dataset stored alongside this one busts this node's cache too".
* `ModelSaverNode` writes `./data/models/model_weights.pt` — inside that tree — and **refuses to write anywhere else**: `model_saver_node.py:78` raises `ValueError("Output path must be within the project data directory")`.

Verified directly:

```
fingerprint A: {'exists': True, 'count': 14, 'total_size': 68239450, ...} hash=3f7bca63ef4065bbf6f655f536a18b60
(write ./data/models/model_weights.pt)
fingerprint B: {'exists': True, 'count': 14, 'total_size': 66549149, ...} hash=93e900bc20267d82ea744f018d40128d
=> fingerprint changed by a write under ./data: True
```

So every run the saver changes the weights file, the dataset's fingerprint changes, the dataset misses, and the miss propagates to `TrainingLoop`. **Accidental, not structural** — exactly the kind of reasoning #143 rejected, just pointing the other way this time.

### The bug is still real, and still reachable with production nodes

Take the shipped graph, expand it, and change **one thing**: drop the `ModelSaver` (and its two `Print` consumers). Everything else is untouched and 100% production nodes — real `SequentialModel`, real `Dataset` (MNIST), real `DataLoader`, real `Optimizer`, real `Loss`, real `TrainingLoop`:

```
nodes: ['start-1', 'model-builder', 'train-pipeline__dataset', 'train-pipeline__dataloader',
        'train-pipeline__loss', 'train-pipeline__optimizer', 'train-pipeline__train_loop', 'visualize-loss']

RUN 1: TrainingLoop.execute calls = 1   dataset=completed  dataloader=completed  optimizer=completed  loss=completed  train_loop=completed
RUN 2: TrainingLoop.execute calls = 0   dataset=cached     dataloader=cached     optimizer=cached     loss=cached     train_loop=cached
RUN 3: TrainingLoop.execute calls = 0   dataset=cached     dataloader=cached     optimizer=cached     loss=cached     train_loop=cached

total TrainingLoop.execute calls across 3 runs: 1
```

"Train a model and plot the loss curve, without saving it" is an ordinary teaching graph. Runs 2 and 3 do no training, write no checkpoints, log no metrics, and the canvas reports the preset as `completed`.

### Revised severity and the part that should worry us

The trigger condition is **"nothing writes into the dataset's `data_dir` tree between runs"**, not "any training graph". That is narrower than filed, and the P8 should be read accordingly.

But the protection is fragile in a specific, foreseeable way: `DatasetNode.cache_fingerprint`'s own comment names scoping the walk to just that dataset's files as "a real follow-up". **If that follow-up ever lands, it removes exactly this accidental protection and the shipped CNN-MNIST graph starts hitting this bug.** Anyone picking up that optimisation should land the `TrainingLoop` fix first or at the same time.

### Unchanged from the original filing

* `TrainingLoopNode` still has no `cacheable` declaration and still inherits `True`.
* `SequentialModelNode.define_inputs()` still returns `[]`, so it is still a fully cacheable root that owns weights — still the odd one out among weight-owning nodes.
* The weight-drift measurement (`0.4089 → 0.4042 → 0.3995`) was taken with `TrainingLoopNode.cacheable` **monkeypatched to `False`**, on the hand-built top-level graph, to show what the naive one-line fix would do. It is not from the unmodified code and not from the shipped graph; the original text says "monkeypatched" but it is easy to read past, so: those numbers show `SequentialModel` handing back the same mutated module, i.e. each run continuing training from the last run's weights. That is why the one-line fix is still not correct on its own.

**Comment by treeleaves30760** [comment]
> 中文摘要：這個 bug 在**未經修改的內建教學範例**上就會發生，不需要拿掉任何節點。六個有訓練的內建圖裡有四個沒有任何寫檔節點，所以那個「意外的保護」不存在；而且四個全部都用 preset 包住，所以狀態還會顯示 `completed`。實測 `C2-5/MLP-MNIST-Training`：第一次執行訓練了一次，第二次和第三次**一次都沒有訓練**，而畫面回報的是 completed。

## 這個 issue 現在應該升級。反覆修正之後，結論比原本更嚴重。

修正後的版本說「只有拿掉 ModelSaver 才觸發得到」，並留下一個未解的問題：有沒有內建範例本來就會踩到。我掃了，答案是**大部分都會**。

## 掃描結果：六個有訓練的內建圖，四個沒有任何寫檔節點

判準是這個 issue 自己找出來的保護機制——寫進 `./data` 的節點會弄髒 `DatasetNode.cache_fingerprint`（它遞迴走 `./data`），讓 dataset 快取失效並往下游傳播。沒有寫檔節點，就沒有這個保護。

```
6 shipped graphs contain training
  WITH a ./data writer (accidentally protected): 2
  WITHOUT any writer  (would hit the bug):       4

    examples/Usage_Example/GPT-Mini/TrainGPT-Mini/graph.json
    examples/Usage_Example/ResNet-CIFAR10/TrainResNet-CIFAR10/graph.json
    plugins/deep/examples/C3-1/LeNet-MNIST-Training/graph.json
    plugins/foundations/examples/C2-5/MLP-MNIST-Training/graph.json
```

後面兩個是**教學插件的範例**——C2-5 和 C3-1，學生上課直接打開的那些。

## 而且四個全部都用 preset 包住，所以狀態會騙人

```
  preset-wrapped  GPT-Mini/TrainGPT-Mini
  preset-wrapped  ResNet-CIFAR10/TrainResNet-CIFAR10
  preset-wrapped  C3-1/LeNet-MNIST-Training
  preset-wrapped  C2-5/MLP-MNIST-Training
```

這一點在 `graph_engine.py:1858-1865` 確認過：

```python
if status in ("completed", "cached", "skipped"):
    container_done[container_id] += 1
    ...
    if container_done[container_id] >= container_total[container_id]:
        await _maybe_await(on_progress(container_id, "completed", None))
```

`cached` 和 `completed` 加的是同一個計數器。**內部節點全部命中快取的 preset，對外回報 `completed`。**

## 實測，未經修改的內建教學範例

計算 `TrainingLoopNode.execute()` 的真實呼叫次數（不看狀態，因為狀態就是會騙人的那一環）：

```
graph: plugins/foundations/examples/C2-5/MLP-MNIST-Training/graph.json
node types: ['Print', 'SequentialModel', 'Start', 'Visualize', 'preset:Training Pipeline']
  run 1: TrainingLoop.execute called 1 time(s); reported statuses = ['completed']
  run 2: TrainingLoop.execute called 0 time(s); reported statuses = ['cached', 'completed']
  run 3: TrainingLoop.execute called 0 time(s); reported statuses = ['cached', 'completed']
```

沒有拿掉任何節點，沒有改任何接線，就是 repo 裡那張圖。

## 為什麼這在教學上特別糟

學生調一個超參數、按 Run、看到 completed、看到 loss 曲線——曲線是上一次的。他會得出「這個參數沒有影響」的結論，而那個結論是錯的，而且他沒有任何線索知道自己被騙了。

這正是這一波在追的那一類問題：**不報錯、畫面正常、結論錯誤**。和 #244 同一家族，但這個更嚴重，因為連「跑了」這件事本身都是假的。

## 兩個建議

1. **升級優先度。** 它現在標 P8，但它在四個內建範例上、其中兩個是教學插件的、而且完全靜默。
2. **preset 的狀態遮蔽應該分開處理。** 就算 #253 修好了，「preset 內部全命中快取卻回報 completed」對任何 preset 都成立，不只是訓練。使用者看不出 preset 到底跑了沒有。這值得自己一個 issue。

（#259 記錄的耦合仍然成立，而且現在更要緊：那個保護只存在於六分之二的圖上。）

**Comment by treeleaves30760** [comment]
修正後的最終確認，兩個沒有寫檔保護的教學範例都跑過了（數真實的 `execute()` 呼叫，不看狀態）：

```
plugins/foundations/examples/C2-5/MLP-MNIST-Training
  修正前: 1, 0, 0      修正後: 1, 1, 1

plugins/deep/examples/C3-1/LeNet-MNIST-Training
  修正前: 1, 0, 0      修正後: 1, 1, 1
```

兩張圖都未經修改，直接從 repo 讀。

注意第二、三次的回報狀態仍然含有 `cached`——那是最上層的 `SequentialModel`（它現在快取的是結構，不是權重，這是正確的）。preset 仍然回報 `completed` 而不區分內部到底跑了沒有，那是 **#260**，還開著。
